File: algorithm-design/hackerrank/cavity_map/cavity_map.go
package cavitymap

// https://www.hackerrank.com/challenges/cavity-map

import (
	"fmt"
)

// CavityMap - implements the solution to the problem
func CavityMap(grid []string) []string {
	fmt.Printf("%v\n", grid)
	deltas := [][]int{{-1, 0}, {0, -1}, {1, 0}, {0, 1}}
	n := len(grid)
	cells := make([][]rune, n)
	for i := 0; i < n; i++ {
		cells[i] = []rune(grid[i])
	}
	for i := 1; i < n-1; i++ {
		for j := 1; j < n-1; j++ {
			deeper := true
			for _, delta := range deltas {
				if uint16(cells[i+delta[0]][j+delta[1]]) >= uint16(cells[i][j]) {
					deeper = false
				}
			}
			if deeper {
				cells[i][j] = 'X'
			}
		}
	}
	for i := 0; i < n; i++ {
		grid[i] = string(cells[i])
	}
	return grid
}

File: algorithm-design/hackerrank/cavity_map/cavity_map.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/cavity-map

import os
import unittest

from typing import List


def cavity_map(grid: List[str]) -> List[str]:
    deltas = [[-1, 0], [0, -1], [1, 0], [0, 1]]
    n = len(grid)
    cells = [list(line) for line in grid]
    for i in range(1, n - 1):
        for j in range(1, n - 1):
            deeper = True
            for delta in deltas:
                if cells[i + delta[0]][j + delta[1]] >= cells[i][j]:
                    deeper = False
            if deeper:
                cells[i][j] = 'X'
    return [''.join(line) for line in cells]


class TestCode(unittest.TestCase):
    def runner(self, name):
        io_lines = [[[]]] * 2
        for index, template in enumerate(['input%s.txt', 'output%s.txt']):
            path = template % name
            path = os.path.join(os.path.split(__file__)[0], path)
            with open(path, 'r') as handle:
                lines = handle.readlines()
            lines = [line.strip() for line in lines]
            io_lines[index] = lines
        grid = io_lines[0][1:]
        grid = cavity_map(grid)
        self.assertEqual(io_lines[1], grid)

    def test_example(self):
        self.runner('_example')
